addons/zoom-slider/userscript.js

```diff
@@ -0,0 +1,79 @@
+export default async function ({ addon, msg }) {
+  const addSlider = async () => {
+    if (addon.tab.editorMode !== "editor" || addon.self.disabled) {
+      return;
+    }
+
+    const zoom = await addon.tab.waitForElement(".blocklyZoom");
+    const blockly = await addon.tab.traps.getBlockly();
+    window.blockly = blockly;
+
+    const foreignObject = document.createElementNS("http://www.w3.org/2000/svg", "foreignObject");
+    foreignObject.setAttribute("height", "500");
+    foreignObject.setAttribute("width", "120");
+    foreignObject.setAttribute("x", "-126");
+    foreignObject.setAttribute("y", "79");
+    const wrapperDiv = document.createElement("div");
+    wrapperDiv.classList.add("sa-zoom-slider-wrapper");
+    const slider = document.createElement("input");
+    slider.type = "range";
+    slider.min = blockly.mainWorkspace.options.zoomOptions.minScale;
+    slider.max = blockly.mainWorkspace.options.zoomOptions.maxScale;
+    slider.step = "0.01";
+    slider.value = blockly.mainWorkspace.scale;
+    slider.classList.add("sa-zoom-slider");
+    const count = document.createElement("input");
+    count.className = addon.tab.scratchClass("input_input-form", "input_input-small", {
+      others: "sa-zoom-slider-count",
+    });
+    count.value = percentScale(blockly.mainWorkspace.scale);
+    wrapperDiv.append(count, slider);
+    foreignObject.append(wrapperDiv);
+    zoom.append(foreignObject);
+
+    slider.addEventListener("mousedown", (e) => {
+      e.stopPropagation();
+    });
+    slider.addEventListener("input", async () => {
+      blockly.mainWorkspace.setScale(slider.value);
+      if (!countInput) {
+        count.value = percentScale(slider.value);
+      }
+    });
+    let countInput = false;
+    count.addEventListener("input", async () => (countInput = true));
+    count.addEventListener("blur", async () => (countInput = false));
+    count.addEventListener("change", async () => {
+      const newScale = parseFloat(count.value / 100);
+      const { maxScale, minScale } = blockly.mainWorkspace.options.zoomOptions;
+      if (newScale > maxScale) {
+        count.value = maxScale * 100;
+      }
+      if (newScale < minScale) {
+        count.value = minScale * 100;
+      }
+      slider.value = count.value / 100;
+      slider.dispatchEvent(new Event("input"));
+      countInput = false;
+    });
+
+    setInterval(() => {
+      slider.min = blockly.mainWorkspace.options.zoomOptions.minScale;
+      slider.max = blockly.mainWorkspace.options.zoomOptions.maxScale;
+      if (parseFloat(slider.value) !== blockly.mainWorkspace.scale) {
+        slider.value = blockly.mainWorkspace.scale;
+        slider.dispatchEvent(new Event("input"));
+      }
+    }, 200);
+
+    addon.self.addEventListener("disabled", () => {
+      foreignObject.remove();
+    });
+  };
+
+  const percentScale = (scale) => Math.round(parseFloat(scale) * 100);
+
+  addSlider();
+  addon.tab.addEventListener("urlChange", addSlider);
```

Comment:
This will also try to add the slider to the project. The easiest way to fix this is probably to put the `addon.tab.waitForElement` in a `while (true)` loop like other editor addons.

Reply:
It won't add it to the project page, though:
https://github.com/mybearworld/fork-ScratchAddons/blob/06d9d6564eed83f8a17a6300b2569477c53fea7a/addons/zoom-slider/userscript.js#L3-L5
It does probably make sense to make it have the same pattern as other addons, though.

Reply:
Right, I'm not sure if it's getting removed properly though because of these errors:
![image](https://github.com/ScratchAddons/ScratchAddons/assets/81489795/364a529e-7064-46ee-9b92-4d4778e5c1d2)

Reply:
I don't see these errors. What did you do to get them?

Reply:
Switch between the editor and project page.

Reply:
Weird, I don't get these errors when doing that. What browser are you using where that occurs?

Reply:
A recent commit fixed it.

Reply:
Oh, that's good